src/mds/MetricAggregator.cc

```diff
@@ -0,0 +1,342 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "MetricAggregator.h"
+#include "mgr/MgrClient.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_mds
+#undef dout_prefix
+#define dout_prefix *_dout << "mds.metric.aggregator" << " " << __func__
+
+MetricAggregator::MetricAggregator(CephContext *cct, MDSRank *mds, MgrClient *mgrc)
+  : Dispatcher(cct),
+    mgrc(mgrc),
+    mds_pinger(mds) {
+}
+
+void MetricAggregator::ping_all_active_ranks(const std::set<mds_rank_t> &active_ranks) {
+  dout(10) << ": pinging " << active.size() << " active mds(s)" << dendl;
+
+  for (auto &rank : active_ranks) {
+    dout(20) << ": pinging rank=" << rank << dendl;
+    mds_pinger.send_ping(rank);
+  }
+}
+
+int MetricAggregator::init() {
+  dout(10) << dendl;
+
+  pinger = std::thread([this]() {
+      std::unique_lock locker(lock);
+      while (!stopping) {
+        std::set<mds_rank_t> active_ranks(active);
+        // drop this lock as mds pinger would acquire mds_lock
+        locker.unlock();
+        ping_all_active_ranks(active_ranks);
+        double timo = g_conf().get_val<std::chrono::seconds>("mds_ping_interval").count();
+        sleep(timo);
+        locker.lock();
+      }
+    });
+
+  mgrc->set_perf_metric_query_cb(
+    [this](const ConfigPayload &config_payload) {
+      set_perf_queries(config_payload);
+    },
+    [this]() {
+      return get_perf_reports();
+    });
+
+  return 0;
+}
+
+void MetricAggregator::shutdown() {
+  dout(10) << dendl;
+
+  {
+    std::scoped_lock locker(lock);
+    ceph_assert(!stopping);
+    stopping = true;
+  }
+
+  if (pinger.joinable()) {
+    pinger.join();
```

Comment:
`pinger` needs to acquire `mds_lock` in `update_rank0` causing deadlock here. See

````
2020-02-15T03:18:43.348+0000 16206700 10 shutdown: mds.metrics
2020-02-15T03:18:43.531+0000 1ea16700 20 update_rank0: mds.metrics
2020-02-15T03:18:43.531+0000 1ea16700 20 update_rank0: mds.metrics: sending metric updates for 0 clients to rank 0 with sequence number 348, last updated sequence number 862
2020-02-15T03:18:44.156+0000 1f217700 10 mds.metric.aggregator ping_all_active_ranks: pinging 1 active mds(s)
2020-02-15T03:18:44.156+0000 1f217700 20 mds.metric.aggregator ping_all_active_ranks: pinging rank=0
2020-02-15T03:18:44.156+0000 1f217700 10 mds.pinger send_ping: rank=0
2020-02-15T03:18:44.156+0000 1f217700 10 mds.pinger send_ping: sending ping with sequence=349 to rank=0
````

From: /ceph/teuthology-archive/pdonnell-2020-02-15_01:51:13-fs-wip-pdonnell-testing-20200214.224251-distro-basic-smithi/4764868/remote/smithi177/log/ceph-mds.b.log.gz

Reply:
crap -- lock ordering in shutdown sequence.

Reply:
`MetricAggregator::shutdown()` and `MetricsHandler::shutdown()` would need to drop `mds_lock` before calling `Thread::join()`.

@batrick -- dropping the lock seems to be ok -- `MDSDaemon::suicide()` and `MDSRankDispatcher::shutdown()` sets `stopping` flag as a guard. Thoughts?

Reply:
You can use `std::scoped_lock` and `try_lock`. Check stopping flag in `update_rank0` if the lock fails.

Or a `std::queue<std::pair<Message, rank_t>>` of a messages to drain in some tick thread in `MDSRank` holding the `mds_lock`.

Reply:
(I don't recommend dropping the `mds_lock` in shutdown.)

Reply:
> (I don't recommend dropping the `mds_lock` in shutdown.)

Any particular reason for that? MDLog::shutdown() drops mds_lock and waits for its thread to terminate.

Reply:
(FWIW, MDCache does the same drop and reacquire mds_lock trickery)

Reply:
Hmm, okay you're right. Dropping the lock in this case seems acceptable.